cookbook_artifacts_download.go

```diff
@@ -0,0 +1,42 @@
+package chef
+
+import (
+	"fmt"
+	"path"
+)
+
+// DownloadTo downloads a cookbook artifact to the specified local directory on disk
+func (c *CBAService) DownloadTo(name, id, localDir string) error {
+
+	cba, err := c.GetVersion(name, id)
+	if err != nil {
+		return err
+	}
+
+	debug("Downloading %s cookbook artifact with id %s\n", cba.Name, cba.Identifier)
+
+	cookbookService := CookbookService{client: c.client}
+	cookbookLongName := fmt.Sprintf("%v-%v", name, cba.Identifier[0:20])
+	cookbookPath := path.Join(localDir, cookbookLongName)
+
+	downloadErrs := []error{
+		cookbookService.downloadCookbookItems(cba.RootFiles, "root_files", cookbookPath),
+		cookbookService.downloadCookbookItems(cba.Files, "files", path.Join(cookbookPath, "files")),
+		cookbookService.downloadCookbookItems(cba.Templates, "templates", path.Join(cookbookPath, "templates")),
+		cookbookService.downloadCookbookItems(cba.Attributes, "attributes", path.Join(cookbookPath, "attributes")),
+		cookbookService.downloadCookbookItems(cba.Recipes, "recipes", path.Join(cookbookPath, "recipes")),
+		cookbookService.downloadCookbookItems(cba.Definitions, "definitions", path.Join(cookbookPath, "definitions")),
+		cookbookService.downloadCookbookItems(cba.Libraries, "libraries", path.Join(cookbookPath, "libraries")),
+		cookbookService.downloadCookbookItems(cba.Providers, "providers", path.Join(cookbookPath, "providers")),
+		cookbookService.downloadCookbookItems(cba.Resources, "resources", path.Join(cookbookPath, "resources")),
+	}
+
+	for _, err := range downloadErrs {
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
One small thought is that it seems a little inefficient to try to download everything when the error handling is going to return only the first error. Is there a graceful way to check for errors after each piece?

Reply:
It looks like this follows the pattern from the [original](https://github.com/chef/go-chef/blob/master/cookbook_download.go#L44). 

`downloadCookbookItems` itself will also has many possible failure points; however it will stop and return at the first it encounters. 

If we want to continue in spite of failures, we should report all errors back to the caller at this level.  We're in an odd in-between place here though.  This API presents the entire cookbook download as a unit. It seems in that case it should be returning a pass-fail result -- in which case it should stop attempting and return the failure. 

Separately, reading through this and `cookbook_download` again has me thinking we should have a [retry here](https://github.com/chef/go-chef/blob/master/cookbook_download.go#L107) for transient failures

Reply:
Created go-chef #24 to discuss / resolve this.